source/geod24/RingBuffer.d

```diff
@@ -0,0 +1,127 @@
+module geod24.RingBuffer;
+
+import std.range;
+
+/*******************************************************************************
+
+    A ring buffer implementation with fixed size
+
+    Params:
+        T = Type of elements
+
+*******************************************************************************/
+
+class RingBuffer (T)
+{
+
+    /*******************************************************************************
+
+        Params:
+            size = Maximum number of elements
+
+    *******************************************************************************/
+
+    this (size_t size) @safe pure nothrow
+    {
+        this.storage = new T[size];
+        this.storage.length = size;
+        if (size)
+        {
+            this.write_range = cycle(this.storage[]);
+            this.read_range = cycle(this.storage[]);
+        }
+    }
+
+    /// Returns: true if full
+    bool full () @safe @nogc pure nothrow
+    {
+        return this.len == this.storage.length;
+    }
+
+    /// Returns: true if empty
+    bool empty () @safe @nogc pure nothrow
+    {
+        return this.len == 0;
+    }
+
+    /// Returns: Number of elements currently in the buffer
+    size_t length () @safe @nogc pure nothrow
+    {
+        return this.len;
+    }
+
+    /*******************************************************************************
+
+        Insert an element to the buffer
+
+        Params:
+            val = new element of type T
+
+    *******************************************************************************/
+
+    void insert () (auto ref T val) @safe @nogc pure nothrow
+    {
+        assert(!this.full());
+
+        write_range.front() = val;
+        write_range.popFront();
+        this.len++;
+    }
+
+    /// Returns: Element in the front of the buffer
+    T popFront () @safe @nogc pure nothrow
+    {
+        assert(!this.empty());
+        scope (exit) read_range.popFront();
+        this.len--;
+
+        return read_range.front();
+    }
+
+    /// Current number of elements
+    private size_t len;
+
+    /// Underlying storage
+    private T[] storage;
+
+    /// Cyclic range for writes
+    private Cycle!(T[]) write_range;
+
+    /// Cyclic range for reads
+    private Cycle!(T[]) read_range;
+}
+
+// A buffer with 0 size should be both empty and full
+unittest
+{
+    auto buffer = new RingBuffer!int(0);
+    assert(buffer.full());
+    assert(buffer.empty());
+}
+
+unittest
```

Comment:
Does it also work when you add elements, remove, and then add again? Specifically I'm wondering how the `read_range`  behaves.

Reply:
Added a loop to test that case